File: frontend/scripts/prod-healthcheck.mjs
#!/usr/bin/env node
/**
 * prod-healthcheck.mjs — LIVE synthetic health-check for bonbox.dk.
 *
 * Catches the class of breakage where Vercel serves an INCONSISTENT deploy:
 * the served index references a chunk hash that either (a) 404s, or (b) 200s
 * but was built against a DIFFERENT graph, so a shared chunk it destructures
 * no longer exports the name it imports. At runtime that is a module-link
 * `SyntaxError: does not provide an export named 'X'` → the lazy import rejects
 * → App.jsx lazyRetry gives up → the top-level ErrorBoundary shows
 * "Something went wrong / connection issue". A plain 200-check MISSES (b) —
 * every chunk returns 200. This walks the chunk GRAPH and verifies the
 * import/export contract, which is the actual failure surface.
 *
 * Run:  npm run healthcheck:prod
 *       node scripts/prod-healthcheck.mjs --base=https://www.bonbox.dk --json
 * Exit: 0 = healthy, 1 = broken deploy, 2 = script/network error.
 *
 * No deps. Node >= 18 (global fetch). Verifies PROD, never localhost.
 */

const argv = Object.fromEntries(
  process.argv.slice(2).map((a) => {
    const [k, v] = a.replace(/^--/, "").split("=");
    return [k, v ?? true];
  }),
);
const BASE = (argv.base || process.env.HEALTHCHECK_BASE || "https://www.bonbox.dk").replace(/\/$/, "");
const JSON_OUT = !!argv.json;
const EXPECT_HASH = argv["expect-index"]; // optional: pin to the hash the current build emitted (staleness gate)

// Route → the source module its lazy import pulls. We assert the served entry
// chunk's dynamic-import map contains a chunk whose basename starts with this.
const KEY_ROUTES = [
  { route: "/reservations", chunkPrefix: "ReservationsPage-" },
  { route: "/dashboard", chunkPrefix: "DashboardPage-" },
  { route: "/gavekort", chunkPrefix: "GavekortPage-" },
  { route: "/r/:slug", chunkPrefix: "ReservationPublicPage-" },
];

const CHUNK_RE = /(?:assets\/)?([A-Za-z0-9_$-]+-[A-Za-z0-9_$-]{8})\.js/g;
const TIMEOUT_MS = 12000;

async function get(url) {
  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), TIMEOUT_MS);
  try {
    const res = await fetch(url, { signal: ctrl.signal, redirect: "follow", headers: { "cache-control": "no-cache" } });
    const body = res.ok ? await res.text() : "";
    return { status: res.status, body, url: res.url };
  } finally {
    clearTimeout(t);
  }
}

/** Parse a rolldown/rollup chunk into its static import edges + export names. */
function parseChunk(src) {
  // imports: import{a as b,c}from"./Chunk-hash.js";  and side-effect import"./x.js"
  const imports = []; // { file, names:[localImportedName] }
  const importFrom = /import\s*\{([^}]*)\}\s*from\s*["']\.\/([A-Za-z0-9_$-]+-[A-Za-z0-9_$-]{8}\.js)["']/g;
  let m;
  while ((m = importFrom.exec(src))) {
    const names = m[1]
      .split(",")
      .map((s) => s.trim())
      .filter(Boolean)
      .map((s) => (s.includes(" as ") ? s.split(" as ")[0].trim() : s.trim())); // we need the EXPORTED name = left side
    imports.push({ file: m[2], names });
  }
  const bareImport = /import\s*["']\.\/([A-Za-z0-9_$-]+-[A-Za-z0-9_$-]{8}\.js)["']/g;
  while ((m = bareImport.exec(src))) imports.push({ file: m[1], names: [] });

  // exports: export{a,b as c,d as default};  (rolldown emits a single/few export{} blocks)
  const exports = new Set();
  const exportBlock = /export\s*\{([^}]*)\}/g;
  while ((m = exportBlock.exec(src))) {
    for (const part of m[1].split(",")) {
      const p = part.trim();
      if (!p) continue;
      // `x as y` re-exports y; `x` exports x
      exports.add(p.includes(" as ") ? p.split(" as ")[1].trim() : p);
    }
  }
  return { imports, exports };
}

async function main() {
  const errors = [];
  const warnings = [];
  const notes = [];

  // 1. Fetch the HTML shell, extract the entry chunk (<script type=module src=...>).
  const idx = await get(BASE + "/");
  if (idx.status !== 200) {
    errors.push(`index HTML ${idx.status} at ${BASE}/`);
    return finish({ errors, warnings, notes });
  }
  const entryMatch = idx.body.match(/<script[^>]+type="module"[^>]+src="\/assets\/([A-Za-z0-9_$-]+-[A-Za-z0-9_$-]{8})\.js"/)
    || idx.body.match(/src="\/assets\/(index-[A-Za-z0-9_$-]{8})\.js"/);
  if (!entryMatch) {
    errors.push("no module entry <script src=/assets/index-*.js> found in HTML shell");
    return finish({ errors, warnings, notes });
  }
  const entryName = entryMatch[1] + ".js";
  notes.push(`entry=${entryName}`);

  if (EXPECT_HASH && entryName !== EXPECT_HASH && `${EXPECT_HASH}.js` !== entryName) {
    // Staleness gate: the live entry is not what THIS build emitted. Vercel didn't
    // ship the current source. Warn (not fail) unless --strict-stale.
    (argv["strict-stale"] ? errors : warnings).push(
      `STALE deploy: live entry ${entryName} != expected ${EXPECT_HASH}`,
    );
  }

  // 2. Fetch the entry chunk = the dynamic-import map (all route chunks resolve here).
  const entry = await get(`${BASE}/assets/${entryName}`);
  if (entry.status !== 200) {
    errors.push(`entry chunk ${entryName} → ${entry.status}`);
    return finish({ errors, warnings, notes });
  }

  // 3. Route-level presence: each key route's chunk must appear in the entry map + return 200.
  const routeChunkByPrefix = {};
  const routeReport = [];
  for (const { route, chunkPrefix } of KEY_ROUTES) {
    const re = new RegExp(`(?:assets\\/)?(${chunkPrefix.replace(/[-]/g, "\\-")}[A-Za-z0-9_$-]{8})\\.js`);
    const hit = entry.body.match(re);
    if (!hit) {
      errors.push(`route ${route}: no chunk matching ${chunkPrefix}* referenced by entry`);
      routeReport.push({ route, chunk: null, status: null, ok: false });
      continue;
    }
    const file = hit[1] + ".js";
    routeChunkByPrefix[chunkPrefix] = file;
    const r = await get(`${BASE}/assets/${file}`);
    const ok = r.status === 200;
    if (!ok) errors.push(`route ${route}: chunk ${file} → ${r.status}`);
    routeReport.push({ route, chunk: file, status: r.status, ok });
  }
  notes.push({ routes: routeReport });

  // 4. Chunk-GRAPH consistency (the real check). BFS from the entry + every route chunk.
  //    For each chunk: every static-imported chunk must (a) return 200 and
  //    (b) EXPORT every name this chunk imports from it. A missing export = the
  //    module-link SyntaxError that crashes the lazy route in prod.
  const cache = new Map(); // file -> {status, parsed}
  async function load(file) {
    if (cache.has(file)) return cache.get(file);
    const r = await get(`${BASE}/assets/${file}`);
    const parsed = r.status === 200 ? parseChunk(r.body) : null;
    const rec = { status: r.status, parsed };
    cache.set(file, rec);
    return rec;
  }

  const roots = [entryName, ...Object.values(routeChunkByPrefix)];
  const seen = new Set();
  const queue = [...roots];
  let edgesChecked = 0;
  while (queue.length) {
    const file = queue.shift();
    if (seen.has(file)) continue;
    seen.add(file);
    const rec = await load(file);
    if (rec.status !== 200) {
      errors.push(`chunk ${file} → ${rec.status} (referenced in graph)`);
      continue;
    }
    for (const imp of rec.parsed.imports) {
      const dep = await load(imp.file);
      if (dep.status !== 200) {
        errors.push(`${file} imports ${imp.file} which → ${dep.status} (broken chunk link)`);
        continue;
      }
      // THE discriminating signal: does the dep actually export what we import?
      for (const name of imp.names) {
        edgesChecked++;
        if (!dep.parsed.exports.has(name)) {
          errors.push(
            `GRAPH INCONSISTENCY: ${file} imports {${name}} from ${imp.file}, ` +
              `but ${imp.file} does not export "${name}" (exports: ${[...dep.parsed.exports].sort().join(",")}). ` +
              `→ module-link SyntaxError → lazy route rejects → ErrorBoundary.`,
          );
        }
      }
      if (!seen.has(imp.file)) queue.push(imp.file);
    }
  }
  notes.push(`graph: ${seen.size} chunks, ${edgesChecked} import-name edges verified`);

  // 5. Backend liveness (App.jsx pings HEAD /api/health). Fail-soft — a backend
  //    blip is a different alert; we only NOTE it so a frontend gate isn't held
  //    hostage by a Render cold start.
  const apiBase =
    argv["api-base"] ||
    process.env.HEALTHCHECK_API_BASE ||
    BASE.replace(/^https?:\/\/(www\.)?/, "https://api.");
  try {
    const h = await get(`${apiBase}/api/health`);
    notes.push(`backend ${apiBase}/api/health → ${h.status}`);
    if (h.status >= 500) warnings.push(`backend /api/health → ${h.status}`);
  } catch (e) {
    warnings.push(`backend health probe failed: ${e.message}`);
  }

  return finish({ errors, warnings, notes });
}

function finish({ errors, warnings, notes }) {
  const healthy = errors.length === 0;
  if (JSON_OUT) {
    console.log(JSON.stringify({ base: BASE, healthy, errors, warnings, notes }, null, 2));
  } else {
    console.log(`\nBonBox prod health-check — ${BASE}`);
    for (const n of notes) console.log("  ·", typeof n === "string" ? n : JSON.stringify(n));
    if (warnings.length) {
      console.log("\n  WARN:");
      for (const w of warnings) console.log("   ~", w);
    }
    if (errors.length) {
      console.log("\n  BROKEN:");
      for (const e of errors) console.log("   ✗", e);
    }
    console.log(`\n  ${healthy ? "HEALTHY ✓" : "DEPLOY BROKEN ✗ (" + errors.length + " error(s))"}\n`);
  }
  process.exit(healthy ? 0 : 1);
}

main().catch((e) => {
  console.error("healthcheck crashed:", e?.stack || e);
  process.exit(2); // distinct from a detected breakage
});

File: backend/app/services/prod_healthcheck.py
"""
prod_healthcheck — server-side port of frontend/scripts/prod-healthcheck.mjs.

Fetches the LIVE frontend (default https://www.bonbox.dk), extracts the entry
chunk, and verifies every referenced chunk resolves to JavaScript — NOT the
SPA index.html Vercel serves at HTTP 200 for a missing /assets/*.js. That
fallback-served-HTML is the exact stale-deploy trap a plain 200-check misses,
and it is what dead-ends users at the ErrorBoundary (the /reservations outage).

Sync (httpx.Client) because it runs on APScheduler's BackgroundScheduler
threads. Bounded (≤ _MAX_FETCHES) + fully fail-soft so a 5-min tick is cheap
and can NEVER crash the web process. Read-only: GETs public URLs only, mutates
nothing — it observes, a human acts.
"""
import re
from dataclasses import dataclass, field

import httpx

_TIMEOUT = httpx.Timeout(12.0, connect=6.0)
_MAX_FETCHES = 90  # covers the full ~76-chunk graph; truncation is a NOTE, never an alarm

# route → the source module its lazy import pulls. We assert the entry's
# dynamic-import map references a chunk whose basename starts with this AND
# that chunk resolves to JS. Mirrors prod-healthcheck.mjs KEY_ROUTES.
KEY_ROUTES = [
    ("/reservations", "ReservationsPage-"),
    ("/dashboard", "DashboardPage-"),
    ("/gavekort", "GavekortPage-"),
    ("/r/:slug", "ReservationPublicPage-"),
]

_ENTRY_RE = re.compile(
    r'<script[^>]+type="module"[^>]+src="/assets/([A-Za-z0-9_$-]+-[A-Za-z0-9_$-]{8})\.js"'
)
_ENTRY_RE_FALLBACK = re.compile(r'src="/assets/(index-[A-Za-z0-9_$-]{8})\.js"')
_IMPORT_FROM = re.compile(
    r"""import\s*\{([^}]*)\}\s*from\s*["']\./([A-Za-z0-9_$-]+-[A-Za-z0-9_$-]{8}\.js)["']"""
)
_EXPORT_BLOCK = re.compile(r"export\s*\{([^}]*)\}")


@dataclass
class HealthResult:
    healthy: bool
    entry: str | None = None
    errors: list = field(default_factory=list)
    summary: str = ""
    chunks_checked: int = 0
    truncated: bool = False


def _looks_like_js(status: int, body: str) -> bool:
    """A chunk URL that returns the SPA index.html (Vercel 404→fallback, HTTP
    200) is a MISSING asset. Reject anything that opens like HTML."""
    if status != 200:
        return False
    head = (body or "").lstrip()[:48].lower()
    if head.startswith("<!") or head.startswith("<html") or "<head" in head:
        return False
    return bool(body)


def _parse_chunk(src: str):
    """(imports, exports) for one rolldown/rollup chunk. imports = list of
    (dep_file, [imported_export_names]); exports = set of names it provides."""
    imports = []
    for m in _IMPORT_FROM.finditer(src):
        names = []
        for part in m.group(1).split(","):
            p = part.strip()
            if not p:
                continue
            # `x as y` imports the EXPORTED name x (left side)
            names.append(p.split(" as ")[0].strip() if " as " in p else p)
        imports.append((m.group(2), names))
    exports = set()
    for m in _EXPORT_BLOCK.finditer(src):
        for part in m.group(1).split(","):
            p = part.strip()
            if not p:
                continue
            # `x as y` re-exports y; `x` exports x
            exports.add(p.split(" as ")[1].strip() if " as " in p else p)
    return imports, exports


def check_prod_frontend(base: str = "https://www.bonbox.dk") -> HealthResult:
    """Probe the live frontend. Returns healthy=False (with errors) on a
    broken/stale deploy OR a probe error. NEVER raises."""
    base = (base or "https://www.bonbox.dk").rstrip("/")
    errors: list[str] = []
    entry_name = None
    fetches = 0
    truncated = False

    try:
        with httpx.Client(
            timeout=_TIMEOUT,
            follow_redirects=True,
            headers={"cache-control": "no-cache", "user-agent": "bonbox-monitor/1.0"},
        ) as c:
            # 1) HTML shell → entry chunk
            idx = c.get(base + "/")
            if idx.status_code != 200:
                return HealthResult(False, None, [f"index HTTP {idx.status_code}"],
                                    f"index HTTP {idx.status_code} at {base}/")
            html = idx.text
            m = _ENTRY_RE.search(html) or _ENTRY_RE_FALLBACK.search(html)
            if not m:
                return HealthResult(False, None,
                                    ["no module entry <script src=/assets/index-*.js> in HTML"],
                                    "no entry script in HTML shell")
            entry_name = m.group(1) + ".js"

            cache: dict[str, object] = {}

            def load(fn):
                """Return (body, parsed) | None (broken: 404/HTML) | 'TRUNCATED'
                (out of budget — a note, not a failure)."""
                nonlocal fetches, truncated
                if fn in cache:
                    return cache[fn]
                if fetches >= _MAX_FETCHES:
                    truncated = True
                    return "TRUNCATED"
                r = c.get(f"{base}/assets/{fn}")
                fetches += 1
                if not _looks_like_js(r.status_code, r.text):
                    cache[fn] = None
                    return None
                rec = (r.text, _parse_chunk(r.text))
                cache[fn] = rec
                return rec

            # 2) entry chunk must resolve to JS (the primary discriminator)
            entry_rec = load(entry_name)
            if entry_rec in (None, "TRUNCATED"):
                errors.append(f"entry {entry_name} did not resolve to JavaScript "
                              f"(SPA fallback / 404) — STALE DEPLOY")
                return HealthResult(False, entry_name, errors, f"entry={entry_name} broken", fetches)
            entry_body = entry_rec[0]

            # 3) each key route's chunk must be referenced by the entry + load as JS
            route_chunks = []
            for route, prefix in KEY_ROUTES:
                rm = re.search(rf"(?:assets/)?({re.escape(prefix)}[A-Za-z0-9_$-]{{8}})\.js",
                               entry_body)
                if not rm:
                    errors.append(f"route {route}: no {prefix}* chunk referenced by entry")
                    continue
                fn = rm.group(1) + ".js"
                route_chunks.append(fn)
                rec = load(fn)
                if rec is None:
                    errors.append(f"route {route}: chunk {fn} did not resolve to JS "
                                  f"(404/HTML) — stale deploy")

            # 4) BOUNDED graph — direct import contracts from entry + route chunks
            #    (one level, not a full BFS). Catches the "chunk 200s but was
            #    built against a different graph" case: an import of a name the
            #    dep no longer exports = a module-link error that crashes the route.
            for fn in [entry_name] + route_chunks:
                rec = cache.get(fn)
                if not isinstance(rec, tuple):
                    continue
                # rec = (body, (imports, exports)); imports = [(dep_file, [names])]
                for dep_file, names in rec[1][0]:
                    dep = load(dep_file)
                    if dep == "TRUNCATED":
                        continue
                    if dep is None:
                        errors.append(f"{fn} imports {dep_file} which did not resolve to JS "
                                      f"— broken chunk link")
                        continue
                    # dep = (body, (imports, exports)); exports set is dep[1][1]
                    missing = [n for n in names if n not in dep[1][1]]
                    if missing:
                        errors.append(f"GRAPH: {fn} imports {{{','.join(missing)}}} from "
                                      f"{dep_file} but it isn't exported — module-link error")
    except Exception as e:  # DNS/TLS/timeout/parse — inconclusive tick, treated as a soft-fail
        return HealthResult(False, entry_name, [f"probe error: {e}"], f"probe error: {e}", fetches)

    healthy = len(errors) == 0
    if healthy:
        summary = f"entry={entry_name}; {fetches} chunks OK"
        if truncated:
            summary += " (graph walk truncated)"
    else:
        summary = f"entry={entry_name}; {errors[0]}"
    return HealthResult(healthy, entry_name, errors, summary, fetches, truncated)


if __name__ == "__main__":  # one-shot manual check: python -m app.services.prod_healthcheck [base]
    import sys
    r = check_prod_frontend(sys.argv[1] if len(sys.argv) > 1 else "https://www.bonbox.dk")
    print("healthy:", r.healthy)
    print("entry:", r.entry)
    print("summary:", r.summary)
    for e in r.errors:
        print("  ✗", e)
    sys.exit(0 if r.healthy else 1)
